### PR TITLE
Add tests for #4395

### PR DESCRIPTION
**Pull request by leni1**
This pull request tracks the fixes for issue #4395

### CODE REVIEW

**Comment by leni1** [comment]
@6a68 good catch. Thought I had run linting tools on this before pushing but I hadn't :(
All fixed now :smile: 
@g-k Not very sure on how to proceed with verifying that an image is JPEG. A Google search I did turned up this answer: https://stackoverflow.com/a/266731. I don't think you guys would want to introduce another dependency into the project by using the Python Imaging Library or adapting the first half of the answer for the image test. Your thoughts on this are appreciated :)

**Comment by g-k** [comment]
Hey @leni1! Good to see you :wave:

You are right, we don't need to introduce PIL or use `is_jpg` from the stackoverflow answer. The server verifies that the image is a JPEG using a check like the `is_jpg` answer but on a base64 data url:

https://github.com/mozilla-services/screenshots/blob/793434f68c5f964afcbeb25aa72682039a27f76e/server/src/servershot.js#L42-L51

So we basically need to add two tests:

* a JPEG version of `test_invalid_data_image` that should trigger the "invalid image" error from line 45
* a JPEG version of `test_invalid_data_image_decoded` that should trigger the "invalid jpeg" error from line 50

It looks like you have a great start in this PR. :+1: 

Also, if I remember right, you had trouble running the screenshots server, so try testing data urls with this script: https://gist.github.com/g-k/aa53e41a0ce6815d76189e8dead00d77

**Comment by leni1** [comment]
@g-k When I ran the script, the invalid data supplied threw an error and the build is green so I guess my test works :woman_shrugging: 
Let me know if there are any fixes not covered by what I've written here.

**Comment by g-k** [comment]
Thanks @leni1! Changes look good.

I ran the server tests against the server with coverage enabled, but didn't see a report for `servershot.js` so I can't confirm that that this covers the right branches.

I think https://github.com/mozilla-services/screenshots/issues/4728 is blocking us until istanbul-middleware is fixed or we switch to another coverage middleware.

**Comment by 6a68** [comment]
@g-k Re-enabling code coverage for this repo is not a high priority. Is there another way the different error cases could be verified? Is it possible to intercept the server's logging output and parse the error message?

**Comment by g-k** [comment]
@6a68 yeah, definitely. I'll give that a shot.

File: test/server/test_image_validation.py [hunk 1]
```diff
@@ -77,7 +77,61 @@ def test_invalid_data_url():
         assert resp.status_code == 500
 
 
+def test_invalid_data_jpeg_image():
+    image = image_setup()
```
Comment (g-k) [inline test/server/test_image_validation.py]: `image_setup` is returning a data url for a png, which won't work for JPEG tests.

File: test/server/test_image_validation.py [hunk 2]
```diff
@@ -77,7 +77,61 @@ def test_invalid_data_url():
...
+    (shot_data, shot_json, user) = user_setup()
+    valid_header = "data:image/jpeg;base64,/9j/2wBDAAQDAwQDAwQEAwQFBAQ"
+    invalid_header = "data:image/jpeg;base64,BAAQ"
+    if valid_header in image:
```
Comment (g-k) [inline test/server/test_image_validation.py]: This should be an assert, so the test doesn't silently pass, without hitting the server.

File: test/server/test_image_validation.py [hunk 3]
```diff
@@ -77,7 +77,61 @@ def test_invalid_data_url():
...
+        resp = user.session.put(
+            shot_data,
+            json=shot_json,
+        )
+        print(resp.text)
+        assert resp.status_code == 500
+
+
+def invalid_data_jpeg_image_decoded():
+    image = image_setup()
```
Comment (g-k) [inline test/server/test_image_validation.py]: Like above we want a JPEG example image instead.

File: test/server/test_image_validation.py [hunk 4]
```diff
@@ -77,7 +77,61 @@ def test_invalid_data_url():
...
+    (shot_data, shot_json, user) = user_setup()
+    if "2wBDAAQDAwQDAw" in image:
```
Comment (g-k) [inline test/server/test_image_validation.py]: Same as above; change to an assert

File: test/server/test_image_validation.py [hunk 5]
```diff
@@ -77,7 +77,61 @@ def test_invalid_data_url():
         assert resp.status_code == 500
 
 
+def test_invalid_data_jpeg_image():
```
Comment (g-k) [inline test/server/test_image_validation.py]: This test can go away if we aren't calling it in main.

File: test/server/test_image_validation.py [hunk 6]
```diff
@@ -77,7 +77,61 @@ def test_invalid_data_url():
...
+            json=shot_json,
+        )
+        print(resp.text)
+        assert resp.status_code == 500
+
+
+def test_invalid_jpeg_data_url():
+    image = image_setup()
+    (shot_data, shot_json, user) = user_setup()
+    if "data:image/jpeg;base64" in image:
```
Comment (g-k) [inline test/server/test_image_validation.py]: Change if to assert

File: test/server/test_image_validation.py [hunk 7]
```diff
@@ -77,7 +77,61 @@ def test_invalid_data_url():
...
+        resp = user.session.put(
+            shot_data,
+            json=shot_json,
+        )
+        print(resp.text)
+        assert resp.status_code == 500
+
+
+def test_invalid_jpeg_data_url():
+    image = image_setup()
```
Comment (g-k) [inline test/server/test_image_validation.py]: Use a JPEG data url

**Comment by 6a68** [comment]
@leni1 Hi! Been about a month since the last review on this PR. Would you like us to hold it open a while longer, or should we close it for now (can always be reopened when you have time)?

**Comment by 6a68** [comment]
@leni1 Sure, that's fine. We'll leave it open a while longer :-)

**Comment by leni1** [comment]
@g-k hi again. Made the changes you suggested to the `test_image_validation` module. One CircleCI build passes and the other fails. Will try and look at the logs to see what is happening.

**Comment by g-k** [comment]
hey @leni1, thanks!

The test failure is probably unrelated to your changes (looks like the test database fails to start or DB migrations aren't running), so I wouldn't worry about that.

I'll test out your changes locally.

File: test/server/test_image_validation.py [hunk 8]
```diff
@@ -25,59 +19,103 @@ def user_setup():
...
+    for clip_id in shot_json['clips']:
+        shot_json['clips'][clip_id]['image']['url'] = invalid_header
+        break
+
+    resp = user.session.put(shot_data, json=shot_json,)
+    print(resp.text)
+    assert resp.status_code == 500
+
+
+def test_invalid_data_jpeg_image_decoded():
```
Comment (g-k) [inline test/server/test_image_validation.py]: This test is failing.

It returns HTTP status code 200 when we want 500. The original version works though, so we can switch back to that.

File: test/server/test_image_validation.py [hunk 9]
```diff
@@ -25,59 +19,103 @@ def user_setup():
...
+        shot_json['clips'][clip_id]['image']['url'] = invalid_header
+        break
+
+    resp = user.session.put(shot_data, json=shot_json,)
+    print(resp.text)
+    assert resp.status_code == 500
+
+
+def test_invalid_png_data_image_decoded():
+    image = example_images['example_images'][0]['url']
```
Comment (g-k) [inline test/server/test_image_validation.py]: For this test and others I was getting a key error and needed to replace `example_images['example_images']` with `example_images` to fix it.

**Comment by g-k** [comment]
@leni1 Sure, I'll take a look on Monday.

**Comment by leni1** [comment]
So tested out the logic for `test_invalid_data_jpeg_image_decoded` and it seems to work fine on my computer with the following output:
```javascript
[official@trojan screenshots]$ node ./screenshots_img_validation.js 'data:image/jpeg;base64,someIM4gEt0Try'
testing: data:image/jpeg;base64,someIM4gEt0Try
invalid-data-image-decoded { msg: 'Invalid JPEG image (after base64 decoding)' }
/home/official/Documents/github/screenshots/screenshots_img_validation.js:64
      throw new Error("invalid jpeg");
      ^

Error: invalid jpeg
    at assertPngOrJpeg (/home/official/Documents/github/screenshots/screenshots_img_validation.js:64:13)
    at Object.<anonymous> (/home/official/Documents/github/screenshots/screenshots_img_validation.js:70:1)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
    at startup (internal/bootstrap/node.js:279:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:696:3)
```
As noted above however, one CircleCI build is passing and another is failing.

**Comment by g-k** [comment]
@leni1 The CI test failure looks unrelated. Can you try rebasing on master to see if that fixes the issue?

Things look good otherwise.

`test_invalid_data_jpeg_image_decoded` is passing a longer data url than in your example since it does a find and replace (it does `invalid_data_image = image.replace('/9j/2wBDAAQDAwQDAw', 'someIM4gEt0Try')`).

Try changing that to just assigning `data:image/jpeg;base64,someIM4gEt0Try` without the replace. I think that would be clearer (less going on since it's assigning a hardcoded string).

**Comment by leni1** [comment]
Have done a `git pull origin master --rebase --ff --allow-unrelated-histories` and `git push -f leni1` (`leni1` is how I refer to my fork). Everything seems fine. :smile:

File: test/server/test_image_validation.py [hunk 10]
```diff
@@ -25,59 +19,82 @@ def user_setup():
...
+    resp = user.session.put(shot_data, json=shot_json,)
+    print(resp.text)
+    assert resp.status_code == 500
+
+
+def test_invalid_data_jpeg_image_decoded():
+    image = example_images[3]['url']
+    (shot_data, shot_json, user) = user_setup()
+    assert '/9j/2wBDAAQDAwQDAw' in image
+    invalid_data_image = 'data:image/jpeg;base64,someIM4gEt0Try'
```
Comment (g-k) [inline test/server/test_image_validation.py]: :+1:

**Comment by g-k** [comment]
@6a68 do you want to take a look?

**Comment by 6a68** [comment]
Whoops! Review queue got a bit long. Happy to merge. Thanks for your continued contributions, @leni1 👍

**Pull request merged**